openfl-tutorials/experimental/Workflow_Interface_401_FedProx_with_Synthetic_nonIID.ipynb

```diff
@@ -0,0 +1,802 @@
+    "\n",
+    "        # Rebuild the optimizer with global model parameters\n",
+    "        self.optimizer = FedProxOptimizer(\n",
+    "            self.model.parameters(), lr=learning_rate, mu=mu, weight_decay=weight_decay)\n",
+    "        # Set global model parameters as old weights to enable computation of proximal term\n",
+    "        self.optimizer.set_old_weights([p.clone().detach() for p in self.model.parameters()])\n",
+    "\n",
+    "        for epoch in range(local_epoch):\n",
+    "            train_loss = []\n",
+    "            correct = 0\n",
+    "            for data, target in self.train_loader:\n",
+    "                self.optimizer.zero_grad()\n",
+    "                output = self.model(data)\n",
+    "                loss = cross_entropy(output, target)\n",
+    "                loss.backward()\n",
+    "                self.optimizer.step()\n",
+    "                pred = output.argmax(dim=1, keepdim=True)\n",
+    "                tar = target.argmax(dim=1, keepdim=True)\n",
+    "                correct += pred.eq(tar).sum().cpu().numpy()\n",
+    "                train_loss.append(loss.item())\n",
+    "            training_accuracy = float(correct / self.train_dataset_length)\n",
+    "            training_loss = np.mean(train_loss)\n",
+    "            print(\n",
+    "                \"<Collab: {:<5}> | Train Round: {:<5} | Local Epoch: {:<3}: FedProx Optimization Train Loss {:<.6f}, Train Acc: {:<.6f} [{}/{}]\".format(\n",
+    "                    self.input,\n",
+    "                    self.round_number,\n",
+    "                    epoch,\n",
+    "                    training_loss,\n",
+    "                    training_accuracy,\n",
+    "                    correct, \n",
+    "                    len(self.train_loader.dataset)\n",
+    "                )\n",
+    "            )\n",
+    "\n",
+    "        self.next(self.join)\n",
+    "    \n",
+    "    @aggregator\n",
+    "    def join(self, inputs):\n",
+    "        \"\"\"\n",
+    "        Compute train dataset, and take weighted average of model.\n",
+    "        \"\"\"\n",
+    "        train_datasize = sum([input_.train_dataset_length for input_ in inputs])\n",
+    "\n",
+    "        train_weights, model_state_dict_list = [], [] \n",
+    "        for input_ in inputs:\n",
+    "            train_weights.append(input_.train_dataset_length / train_datasize)\n",
+    "            model_state_dict_list.append(input_.model.state_dict())\n",
+    "\n",
+    "        avg_model_dict = weighted_average(model_state_dict_list, train_weights)\n",
+    "        self.model.load_state_dict(avg_model_dict)\n",
+    "\n",
+    "        self.next(self.end)\n",
+    "\n",
+    "    @aggregator\n",
+    "    def end(self):\n",
+    "        \"\"\"\n",
+    "        This is the 'end' step.\n",
+    "        \"\"\"\n",
+    "        self.round_number += 1\n",
+    "        print('This is end of the flow')"
```

Comment:
This print statement is misleading because it gets printed after every round. Let's pass `total_rounds` to federated flow and print this condition once at the end of the flow.

Reply:
Incorporated